node/src/components/storage.rs

```diff
@@ -649,250 +684,311 @@ impl Storage {
-                    .ignore()
+                    .respond(self.get_highest_block_with_metadata().await?)
+                    .await;
             }
             StorageRequest::PutBlockSignatures {
                 signatures,
                 responder,
             } => {
-                let mut txn = self.env.begin_rw_txn()?;
-                let old_data: Option<BlockSignatures> =
-                    txn.get_value(self.block_metadata_db, &signatures.block_hash)?;
-                let new_data = match old_data {
-                    None => signatures,
-                    Some(mut data) => {
-                        for (pk, sig) in signatures.proofs {
-                            data.insert_proof(pk, sig);
-                        }
-                        data
-                    }
-                };
-                let outcome = txn.put_value(
-                    self.block_metadata_db,
-                    &new_data.block_hash,
-                    &new_data,
-                    true,
-                )?;
-                txn.commit()?;
-                responder.respond(outcome).ignore()
+                responder
+                    .respond(self.put_block_signatures(signatures)?)
+                    .await;
             }
             StorageRequest::GetBlockSignatures {
                 block_hash,
                 responder,
             } => {
                 let result =
                     self.get_finality_signatures(&mut self.env.begin_ro_txn()?, &block_hash)?;
-                responder.respond(result).ignore()
+                responder.respond(result).await;
             }
             StorageRequest::GetFinalizedDeploys { ttl, responder } => {
-                responder.respond(self.get_finalized_deploys(ttl)?).ignore()
+                responder.respond(self.get_finalized_deploys(ttl)?).await;
             }
-        })
+        }
+        Ok(())
+    }
+
+    fn put_block_signatures(&self, signatures: BlockSignatures) -> Result<bool, Error> {
+        let mut txn = self.env.begin_rw_txn()?;
+        let old_data: Option<BlockSignatures> =
+            txn.get_value(self.block_metadata_db, &signatures.block_hash)?;
+        let new_data = match old_data {
+            None => signatures,
+            Some(mut data) => {
+                for (pk, sig) in signatures.proofs {
+                    data.insert_proof(pk, sig);
+                }
+                data
+            }
+        };
+        let outcome = txn.put_value(
+            self.block_metadata_db,
+            &new_data.block_hash,
+            &new_data,
+            true,
+        )?;
+        txn.commit()?;
+        Ok(outcome)
+    }
+
+    async fn get_highest_block_with_metadata(
+        &self,
+    ) -> Result<Option<(Block, BlockSignatures)>, Error> {
+        let mut txn = self.env.begin_ro_txn()?;
+        let indices = self.indices.read()?;
+        let maybe_highest_known_height: Option<u64> = {
```

Comment:
`Option<u64>` could probably be inferred by the compiler.

Reply:
That one I like leaving in. Can't hurt to have a little extra type safety!